client/src/components/SendMessage.js

```diff
@@ -36,36 +39,73 @@ class SendMessage extends React.Component {
     return;
   }
 
+  hideEmojiPicker() {
+      this.setState({
+          ...this.state,
+          emojiPickerActive : false
+      });
+  }
+  onKeyUp(event) {
+      if (this.state.emojiPickerActive){
+          if (event.which === 9 || event.which === 37 || event.which === 39) {
+              // Cycle through preview
+              this.refs.emojipicker.onKeyUp(event);
+          } else {
+              const lastWord = this.refs.message.value.split(" ").pop();
+              this.refs.emojipicker.filterWord(lastWord);
+          }
+      }
+  }
+
   onKeyDown(event) {
     // console.log("KEYDOWN", event.type, event.which);
+    if (this.state.emojiPickerActive) {
+        this.refs.emojipicker.onKeyDown(event);
+    }
     if(event.which === 9) {
       // Tab
       event.preventDefault();
       // TODO: autocomplete user names
     } else if(event.which === 186) {
       // ':'
-      // TODO: bring up emoji preview
-    }
+      if (this.props.useEmojis) {
+         this.setState({
+             ...this.state,
+             emojiPickerActive : true
+         });
+      }
+  }
     return;
   }
+  displayEmojiText (emojiText) {
+      let text = this.refs.message.value.split(" ");
+      text.pop();
+      text.push(emojiText);
+      this.refs.message.value = text.join(" ");
+  }
 
   render() {
     return (
       <div className="SendMessage">
         <form onSubmit={this.sendMessage.bind(this)}>
+            {this.props.useEmojis &&
+             this.state.emojiPickerActive ?
+                 <EmojiPicker ref='emojipicker'
+                     hideEmojiPicker={this.hideEmojiPicker.bind(this)}
```

Comment:
I think the hideEmojiPicker would be better renamed to onClose and the callback function this.hideEmojiPicker --> this.onCloseEmojiPicker to better reflect what's happening. Personally I associate "close" to when the component is gone from the DOM whereas "hide" is when the component is in the DOM but not visible.